### DISCUSSION TITLE
How to isolate the individual runs of a hydra multi-run?

### DISCUSSION THREAD

**Discussion by maxfrei750** [facebookresearch/hydra#2186] [Q&A] [answered]
Is there a best practice to isolate the individual runs of a hydra multi-run, i.e. that each run starts fresh and is handled as a separate process entirely? 

I think the [joblib launcher](https://hydra.cc/docs/plugins/joblib_launcher/) might be a way to achieve this behavior, but maybe there is a  simpler way?

Thanks in advance!

**Comment by Jasha10** [comment, marked as answer]
Hi @maxfrei750,
I think Hydra's suite of plugins is currently lacking in this regard. We would need to develop a new launcher that creates a separate process for each job. I've thought about using Python's `multiprocessing` module as the backend for such a purpose -- if my understanding is correct, the `multiprocessing` module would achieve the isolation that you're talking about.

**Reply by Jasha10** [reply]
I've just opened feature request #2187.

**Reply by maxfrei750** [reply]
Hi @Jasha10, thanks for the reply and the hints. I think I should add: For my use-case, I don't need (actually, even don't want) parallelism. Maybe that would make things easier?

**Reply by Jasha10** [reply]
I think using `JoblibLauncher` is the easiest way to go currently. You should be able avoid parallelism by configuring the launcher with `n_jobs` equal to `1`.

**Reply by maxfrei750** [reply]
Unfortunately, I cannot provide a minimum example right now, but I tested `JoblibLauncher` with `n_jobs=1` and while it worked, it did not result in an isolation of the runs.

**Reply by Jasha10** [reply]
I just did some quick experimenting with `joblib`. I found that, when `n_jobs=1`, joblib re-uses the main process instead of launching a new process.
I wonder if there is some way to force `joblib` to use a separate process...

**Reply by maxfrei750** [reply]
From the [joblib documentation](https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html):

> If 1 is given, no parallel computing code is used at all, which is useful for debugging.

**Reply by Jasha10** [reply]
I see. What sweeper are you using? Here's an idea:
- set joblib's `n_jobs` to `2`
- set the sweeper's batch size to `1`.
Using Hydra's `BasicSweeper` this can be done by setting `hydra.sweeper.max_batch_size` to `1`.

My thinking is that setting the sweeper's batch size to `1` would force sequential job execution, and using joblib `n_jobs==2` would enable process isolation.

**Reply by maxfrei750** [reply]
I think that might work. 

_Without_ `hydra.sweeper.max_batch_size=1` I get:

```
[2022-05-02 22:20:57,098][HYDRA] Joblib.Parallel(n_jobs=2,backend=loky,prefer=processes,require=None,verbose=0,timeout=None,pre_dispatch=2*n_jobs,batch_size=auto,temp_folder=None,max_nbytes=None,mmap_mode=r) is launching 3 jobs
[2022-05-02 22:20:57,098][HYDRA] Launching jobs, sweep output dir : output/demo_endoscope_joblib
[2022-05-02 22:20:57,098][HYDRA] 	#0 : initial_runtime_state.seed=1
[2022-05-02 22:20:57,098][HYDRA] 	#1 : initial_runtime_state.seed=2
[2022-05-02 22:20:57,098][HYDRA] 	#2 : initial_runtime_state.seed=3
```

_With_ `hydra.sweeper.max_batch_size=1` I get:

```
[2022-05-02 22:22:39,288][HYDRA] Joblib.Parallel(n_jobs=2,backend=loky,prefer=processes,require=None,verbose=0,timeout=None,pre_dispatch=2*n_jobs,batch_size=auto,temp_folder=None,max_nbytes=None,mmap_mode=r) is launching 1 jobs
[2022-05-02 22:22:39,288][HYDRA] Launching jobs, sweep output dir : output/demo_endoscope_joblib
[2022-05-02 22:22:39,288][HYDRA] 	#0 : initial_runtime_state.seed=1
```

Unfortunately, in both cases, I run into the following error:
```
Exception has occurred: BrokenProcessPool
A task has failed to un-serialize. Please ensure that the arguments of the function are all picklable.
```

But I think that is an entirely different issue.

EDIT: Corrected last paragraph.

**Reply by Jasha10** [reply]
🤔
Yes, that does seem like a separate issue.
I can't help much without having a way to reproduce the issue myself...
I wonder if changing the joblib backend would make a difference (e.g. setting `hydra.launcher.backend=="multiprocessing"`).
It seems that Joblib prefers the `"loky"` backend by default.

**Reply by maxfrei750** [reply]
Sorry for the late reply.

I wonder if changing the joblib backend would make a difference (e.g. setting `hydra.launcher.backend=="multiprocessing"`).

I was surprised that this is even an option, since the [hydra documentation](https://hydra.cc/docs/plugins/joblib_launcher/) says that `"loky"` was the only supported backend. Unfortunately, for my configuration, it resulted in the same un-serialization error.

I'd really like to provide a minimum reproduction, but unfortunately I struggle, without including my whole, rather large, structured config.

I think it would already help quite a lot, to know, at what point of the code the un-serialization takes place. Because I managed for instance to pickle/unpickle my config (raw, as well as instantiated) without any problems. :thinking:

**Reply by maxfrei750** [reply]
I finally managed to get a more detailed error stack:
```
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/joblib/externals/loky/process_executor.py", line 407, in _process_worker
    call_item = call_queue.get(block=True, timeout=timeout)
  File "/usr/local/.pyenv/versions/3.10.2/lib/python3.10/multiprocessing/queues.py", line 122, in get
    return _ForkingPickler.loads(res)
  File "/workspace/synthpic2/__init__.py", line 3, in <module>
    from .engine import execute_recipe
  File "/workspace/synthpic2/engine.py", line 6, in <module>
    import bpy
  File "/opt/bpy/lib/python3.10/site-packages/3.1/scripts/modules/bpy/__init__.py", line 38, in <module>
    from _bpy import (
ModuleNotFoundError: No module named '_bpy'
```

It tells me that the problem lies in the [Blenders](https://www.blender.org/) internal C module (`_bpy`). I'm well aware, that this issue is not even remotely related to hydra. But maybe you still have some helpful insights for me?

**Reply by maxfrei750** [reply]
Related: https://github.com/TylerGubala/blenderpy/issues/23#issuecomment-514826760

**Reply by maxfrei750** [reply]
I solved the pickling problem with the aforementioned link, so that the joblib launcher now works. So this brings us back to the original topic of this question: run isolation :partying_face: 

Unfortunately, the runs are still not fully isolated, but rather seem to be grouped into groups of two. I guess that this is the effect of `n_jobs=2`.

Would it be feasible to write a custom launcher that starts each run in a separate process using `subprocess.Popen` or something similar? Or is that an inherently bad idea?

**Reply by Jasha10** [reply]
> Would it be feasible to write a custom launcher that starts each run in a separate process using `subprocess.Popen` or something similar? Or is that an inherently bad idea?

One requirement on launchers is that they should be able to facilitate communication from the finished job back to the sweeper. Specifically, the job's return value needs to be sent back to the sweeper (wrapped in a JobReturn object).

If the job returns a complex object (e.g. if the main function returns an instance of MyCustomReturnType), it could be hard to send that returned object via a Popen'd pipe. Popen is text-based (i.e. stdout and stderr are IO streams).

**Reply by maxfrei750** [reply]
I see. Thanks for elaborating! Then I guess that the `multiprocessing` module is really the easiest way... just like you wrote in your initial reply.  :sweat_smile: Hence, I think I should finally formally accept your answer and the issue you opened (#2187) is a better place for further discussion. Thank you very much!

**Reply by IlyasMoutawwakil** [reply]
@maxfrei750 @Jasha10, first of all thank you both for your work on this topic.
what makes the two runs not perfectly isolated ? I followed your suggestion:
```yaml
hydra:
  launcher:
    n_jobs: 2
    prefer: processes # use processes instead of threads
    backend: multiprocessing # use multiprocessing backend instead of loky
  sweeper:
    max_batch_size: 1
 ```
I'm trying to achieve perfect isolation between torch or onnxruntime benchmarks. 
It seems that the two runs have different PIDs, doesn't that guarantee isolation ?

**Reply by romesco** [reply]
> I solved the pickling problem with the aforementioned link, so that the joblib launcher now works. So this brings us back to the original topic of this question: run isolation 🥳

Hey @maxfrei750, what was the solution to solving the "failed to un-serialize" issue? How did you temporarily swap the system path problem in when the deserialization happens via the hydra->joblib pipeline?

**Reply by maxfrei750** [reply]
@romesco I'm not sure, if the solution to my problem can be generalized, since it was closely related to Blender's `bpy` module and the way it interacts with the system path (see https://github.com/TylerGubala/blenderpy/issues/23#issuecomment-514826760, if you haven't done that already). If I remember correctly (it's been a while), I simply saved and restored the system `sys.path` variable like this:

```python
ORIGINAL_SYS_PATH = list(sys.path)

import bpy

sys.path = ORIGINAL_SYS_PATH
```

And the problem was solved. But once again: This is probably very specific to the `bpy` library.
